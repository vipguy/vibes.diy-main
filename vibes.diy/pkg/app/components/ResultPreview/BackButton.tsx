import { BackArrowIcon } from "../HeaderContent/SvgIcons.js";
import React from "react";

interface BackButtonProps {
  onBackClick: () => void;
}

export const BackButton: React.FC<BackButtonProps> = ({ onBackClick }) => {
  return (
    <button
      type="button"
      onClick={onBackClick}
      className="bg-light-decorative-00 dark:bg-dark-decorative-00 text-light-primary dark:text-dark-primary hover:bg-light-decorative-01 dark:hover:bg-dark-decorative-01 flex items-center justify-center rounded-md p-2 transition-colors md:hidden"
      aria-label="Back to chat"
    >
      <BackArrowIcon />
    </button>
  );
};
