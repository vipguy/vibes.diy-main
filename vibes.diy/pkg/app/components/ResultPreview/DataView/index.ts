// Export all data view components
export { default as DatabaseData } from "./DatabaseData.js";
export { default as DatabaseListView } from "./DatabaseListView.js";
export { default as DynamicTable } from "./DynamicTable.js";
export { headersForDocs } from "./dynamicTableHelpers.js";
