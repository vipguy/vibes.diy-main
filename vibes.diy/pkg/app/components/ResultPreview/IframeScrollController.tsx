import { useEffect, useRef } from "react";

// Create static refs that persist across component remounts
// These will maintain state even when the component is unmounted and remounted
const staticRefs = {
  scroller: null as HTMLElement | null,
  contentObserver: null as MutationObserver | null,
  checkForScrollerInterval: null as NodeJS.Timeout | null,
  lastScrollHeight: 0,
};

interface SandpackScrollControllerProps {
  isStreaming: boolean;
  shouldEnableScrolling?: boolean;
  codeReady?: boolean;
  activeView?: "preview" | "code";
}

const SandpackScrollController: React.FC<SandpackScrollControllerProps> = ({
  isStreaming,
  // shouldEnableScrolling = isStreaming, // Default to isStreaming if not provided
  codeReady = false,
  activeView = "preview", // Default to preview view
}) => {
  // Keep component-level refs for React's hook rules
  const componentMounted = useRef(false);
  const propsRef = useRef({ isStreaming, codeReady, activeView });

  // Update props ref when they change
  useEffect(() => {
    propsRef.current = { isStreaming, codeReady, activeView };
  }, [isStreaming, codeReady, activeView]);

  // Simple check if we should be auto-scrolling
  const shouldAutoScroll = () => {
    const { isStreaming, codeReady, activeView } = propsRef.current;
    return isStreaming && !codeReady && activeView === "code";
  };

  // Immediately scroll to the bottom with no animations
  const scrollToBottom = () => {
    if (!staticRefs.scroller) return;

    // Hard scroll to bottom with no animation
    staticRefs.scroller.scrollTop = staticRefs.scroller.scrollHeight;
    staticRefs.lastScrollHeight = staticRefs.scroller.scrollHeight;
  };

  // Setup the scroller observer
  const setupScroller = (scroller: HTMLElement) => {
    if (!scroller) return;

    staticRefs.scroller = scroller;

    // Clean up any previous observers
    if (staticRefs.contentObserver) {
      staticRefs.contentObserver.disconnect();
    }

    // Setup content observer with simplified mutation handling
    const contentObserver = new MutationObserver(() => {
      // Skip if component not mounted or no scroller
      if (!componentMounted.current || !staticRefs.scroller) return;

      // Check if content height has changed
      const newHeight = staticRefs.scroller.scrollHeight;
      if (newHeight !== staticRefs.lastScrollHeight && shouldAutoScroll()) {
        scrollToBottom();
      }

      staticRefs.lastScrollHeight = newHeight;
    });

    // Observe all relevant content changes
    contentObserver.observe(scroller, {
      childList: true,
      subtree: true,
      characterData: true,
    });

    staticRefs.contentObserver = contentObserver;

    // Do initial scroll if needed
    if (shouldAutoScroll()) {
      scrollToBottom();
    }
  };

  // Main effect to handle mounting and setup
  useEffect(() => {
    componentMounted.current = true;

    // Add the highlight styles if they don't exist (for test compatibility)
    if (!document.getElementById("highlight-style")) {
      const style = document.createElement("style");
      style.id = "highlight-style";
      style.textContent = `
        .cm-line-highlighted {
          position: relative !important;
          border-left: 3px solid rgba(0, 137, 249, 0.6) !important;
          color: inherit !important;
          transition: border-color 0.4s ease-in-out !important;
        }
        
        .cm-line-highlighted::before {
          content: "" !important;
          position: absolute !important;
          top: 0 !important;
          left: 0 !important;
          right: 0 !important;
          bottom: 0 !important;
          background: linear-gradient(
            90deg, 
            rgba(0, 128, 255, 0.12) 0%, 
            rgba(224, 255, 255, 0.2) 50%, 
            rgba(0, 183, 255, 0.12) 100%
          ) !important;
          background-size: 200% 100% !important;
          pointer-events: none !important;
          z-index: -1 !important;
          opacity: 0 !important;
          transition: opacity 1ms ease-in-out !important;
        }
        
        /* Simple animation for all document sizes */
        .cm-line-highlighted.active::before {
          opacity: 1 !important;
          animation: sparkleFlow 2s ease-in-out infinite !important;
        }
        
        /* Simple animation keyframes */
        @keyframes sparkleFlow {
          0% { background-position: 0% 50%; }
          50% { background-position: 100% 50%; }
          100% { background-position: 0% 50%; }
        }
        
        /* Fade out class with transition */
        .cm-line-fade-out {
          transition: border-color 1.5s cubic-bezier(0.25, 0.1, 0.25, 1) !important;
          border-left-color: rgba(0, 137, 249, 0.2) !important;
        }
        
        .cm-line-fade-out::before {
          opacity: 0 !important;
          transition: opacity 1.5s cubic-bezier(0.25, 0.1, 0.1, 1) !important;
        }
      `;
      document.head.appendChild(style);
    }

    // Find the scroller element
    const findAndSetupScroller = () => {
      const scroller = document.querySelector(".cm-scroller");
      if (scroller && scroller instanceof HTMLElement) {
        setupScroller(scroller);
        return true;
      }
      return false;
    };

    // Try to find scroller immediately
    if (!findAndSetupScroller()) {
      // If not found, set up an interval to keep trying
      staticRefs.checkForScrollerInterval = setInterval(() => {
        if (findAndSetupScroller()) {
          // Once found, clear the interval
          if (staticRefs.checkForScrollerInterval) {
            clearInterval(staticRefs.checkForScrollerInterval);
            staticRefs.checkForScrollerInterval = null;
          }
        }
      }, 100);
    }

    // Cleanup function
    return () => {
      componentMounted.current = false;

      // Clear intervals
      if (staticRefs.checkForScrollerInterval) {
        clearInterval(staticRefs.checkForScrollerInterval);
        staticRefs.checkForScrollerInterval = null;
      }

      // Disconnect observer
      if (staticRefs.contentObserver) {
        staticRefs.contentObserver.disconnect();
      }
    };
  }, []);

  // Effect for responding to prop changes
  useEffect(() => {
    // If conditions change and we should auto-scroll now, do it
    if (shouldAutoScroll() && staticRefs.scroller) {
      scrollToBottom();
    }
  }, [isStreaming, codeReady, activeView]);

  return null;
};

export default SandpackScrollController;
