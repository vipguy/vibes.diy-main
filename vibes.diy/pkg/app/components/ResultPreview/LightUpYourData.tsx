import React from "react";

const LightUpYourData = () => {
  return (
    <div className="flex h-full w-full items-center justify-center">
      <img
        src="/code-things-black.png"
        alt="You can just code things"
        className="logo-pulse opacity-2 dark:hidden"
      />
      <img
        src="/code-things.png"
        alt="You can just code things"
        className="logo-pulse hidden opacity-2 dark:block"
      />
    </div>
  );
};

export default LightUpYourData;
