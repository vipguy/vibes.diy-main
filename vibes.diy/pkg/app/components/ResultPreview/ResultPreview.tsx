import React, { useCallback, useEffect, useMemo } from "react";
import type { RuntimeError } from "@vibes.diy/use-vibes-types";
import { animationStyles } from "./ResultPreviewTemplates.js";
import type { IframeFiles, ResultPreviewProps } from "./ResultPreviewTypes.js";
// import { encodeTitle } from '../SessionSidebar/utils';
// ResultPreview component
import IframeContent from "./IframeContent.js";
import AppSettingsView from "./AppSettingsView.js";
import {
  downloadTextFile,
  generateStandaloneHtml,
} from "../../utils/exportHtml.js";
import { useSession } from "../../hooks/useSession.js";

function ResultPreview({
  code,
  // dependencies,
  onScreenshotCaptured,
  sessionId,
  title,
  updateTitle,
  isStreaming = false,
  codeReady = false,
  displayView,
  onPreviewLoaded,
  setMobilePreviewShown,
  setIsIframeFetching,
  addError,
  children,
  onCodeSave,
  onCodeChange,
  onSyntaxErrorChange,
}: ResultPreviewProps & { children?: React.ReactNode }) {
  // Use CSS-based dark mode detection like the rest of the UI
  const isDarkMode =
    typeof document !== "undefined"
      ? document.documentElement.classList.contains("dark")
      : true; // Default to dark mode for SSR
  const {
    vibeDoc,
    updateDependencies,
    updateInstructionalTextOverride,
    updateDemoDataOverride,
  } = useSession(sessionId);
  const showWelcome = !isStreaming && (!code || code.length === 0);

  // Use title from props directly
  const currentTitle = title || "Untitled App";

  // Settings view callbacks handled in AppSettingsView

  // Calculate filesContent directly based on code prop
  const filesContent = useMemo<IframeFiles>(() => {
    // Always return the expected structure, defaulting code to empty string
    return {
      "/App.jsx": {
        code: code && !showWelcome ? code : "", // Use code if available, else empty string
        active: true,
      },
    };
  }, [code, showWelcome, codeReady, isStreaming]); // Include codeReady to ensure updates

  // Theme is now provided by ThemeContext

  // Function to download HTML file
  const handleDownloadHtml = useCallback(async () => {
    try {
      const html = generateStandaloneHtml({ code });
      const name = currentTitle !== "Untitled App" ? currentTitle : "app";
      downloadTextFile(`${name}.html`, html);
    } catch (error) {
      console.error("Failed to download HTML:", error);
      if (addError) {
        addError({
          type: "error",
          message: "Failed to download HTML file",
          source: "download-html",
          timestamp: new Date().toISOString(),
        });
      }
    }
  }, [code, sessionId, currentTitle, addError]);

  useEffect(() => {
    const handleMessage = ({ data }: MessageEvent) => {
      if (data) {
        if (data.type === "preview-ready" || data.type === "preview-loaded") {
          // No API key needed - proxy handles authentication
          setMobilePreviewShown(true);
          onPreviewLoaded();
        } else if (data.type === "streaming" && data.state !== undefined) {
          if (setIsIframeFetching) {
            setIsIframeFetching(data.state);
          }
        } else if (data.type === "screenshot" && data.data) {
          if (onScreenshotCaptured) {
            onScreenshotCaptured(data.data);
          }
        } else if (data.type === "screenshot-error" && data.error) {
          // Still call onScreenshotCaptured with null to signal that the screenshot failed
          if (onScreenshotCaptured) {
            onScreenshotCaptured(null);
          }
        } else if (data.type === "iframe-error" && data.error) {
          const error = data.error as RuntimeError;
          if (addError) {
            addError(error);
          }
        }
      }
    };
    window.addEventListener("message", handleMessage);
    return () => {
      window.removeEventListener("message", handleMessage);
    };
  }, [
    onScreenshotCaptured,
    onPreviewLoaded,
    setIsIframeFetching,
    setMobilePreviewShown,
    addError,
    sessionId,
    currentTitle,
  ]);

  const previewArea = showWelcome ? (
    <div className="h-full">{/* empty div to prevent layout shift */}</div>
  ) : (
    <>
      <IframeContent
        activeView={displayView}
        filesContent={filesContent}
        isStreaming={!codeReady}
        codeReady={codeReady}
        isDarkMode={isDarkMode}
        sessionId={sessionId}
        onCodeSave={onCodeSave}
        onCodeChange={onCodeChange}
        onSyntaxErrorChange={onSyntaxErrorChange}
      />
      {displayView === "settings" && (
        <div
          style={{
            position: "absolute",
            zIndex: 1,
            height: "100%",
            width: "100%",
            top: 0,
            left: 0,
          }}
        >
          <AppSettingsView
            title={currentTitle}
            onUpdateTitle={updateTitle}
            onDownloadHtml={handleDownloadHtml}
            selectedDependencies={vibeDoc?.dependencies}
            dependenciesUserOverride={vibeDoc?.dependenciesUserOverride}
            aiSelectedDependencies={vibeDoc?.aiSelectedDependencies}
            onUpdateDependencies={updateDependencies}
            instructionalTextOverride={vibeDoc?.instructionalTextOverride}
            demoDataOverride={vibeDoc?.demoDataOverride}
            onUpdateInstructionalTextOverride={updateInstructionalTextOverride}
            onUpdateDemoDataOverride={updateDemoDataOverride}
          />
        </div>
      )}
    </>
  );

  return (
    <div
      className="h-full"
      style={{ overflow: "hidden", position: "relative" }}
    >
      <style>{animationStyles}</style>
      {previewArea}
      {children}
    </div>
  );
}

export default ResultPreview;
