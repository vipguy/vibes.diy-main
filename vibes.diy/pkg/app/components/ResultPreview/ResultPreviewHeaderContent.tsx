import React, { /*useEffect,*/ useRef } from "react"; // useEffect no longer needed here
import { useParams } from "react-router";
import { useSession } from "../../hooks/useSession.js";
import type { ViewControlsType, ViewType } from "@vibes.diy/prompts";
// import { useViewState } from '../../utils/ViewState'; // useViewState is now lifted to home.tsx
import { BackButton } from "./BackButton.js";
import { SaveButton } from "./SaveButton/SaveButton.js";
import { ShareButton } from "./ShareButton.js";
import { ShareModal } from "./ShareModal.js";
import { usePublish } from "./usePublish.js";
import { ViewControls } from "./ViewControls.js";

interface ResultPreviewHeaderContentProps {
  // Props from useViewState (lifted to home.tsx)
  displayView: ViewType;
  navigateToView: (view: ViewType) => void;
  viewControls: ViewControlsType;
  showViewControls: boolean;
  previewReady: boolean;
  setMobilePreviewShown: (shown: boolean) => void;
  setUserClickedBack?: (clicked: boolean) => void;

  // Props required by usePublish and useSession hooks, and for BackButton logic
  code: string; // for usePublish
  isStreaming: boolean; // for BackButton logic
  sessionId?: string; // for useSession, usePublish
  title?: string; // for useSession, usePublish

  // Props for code editing
  hasCodeChanges?: boolean;
  onCodeSave?: () => void;
  syntaxErrorCount?: number;
}

const ResultPreviewHeaderContent: React.FC<ResultPreviewHeaderContentProps> = ({
  displayView,
  navigateToView,
  viewControls,
  showViewControls,
  previewReady,
  setMobilePreviewShown,
  setUserClickedBack,
  code,
  isStreaming,
  sessionId: propSessionId,
  title: propTitle,
  hasCodeChanges,
  onCodeSave,
  syntaxErrorCount,
}) => {
  const { sessionId: urlSessionId, view: urlView } = useParams();
  const publishButtonRef = useRef<HTMLButtonElement>(null);

  // Use props if provided, otherwise use params from the URL
  const sessionId = propSessionId || urlSessionId;
  const title = propTitle || urlView;

  // Use the session hook to get and update session data - only if we have a sessionId
  const {
    session,
    docs: messages,
    updatePublishedUrl,
    updateFirehoseShared,
  } = useSession(sessionId || "temp-session");

  // useViewState is now lifted, props like displayView, navigateToView, viewControls, showViewControls are passed in.
  // The useEffect syncing activeView with displayView is no longer needed.

  // Use the custom hook for publish functionality
  const {
    isPublishing,
    urlCopied,
    publishedAppUrl,
    handlePublish,
    toggleShareModal,
    isShareModalOpen,
    setIsShareModalOpen,
  } = usePublish({
    sessionId,
    code,
    title,
    messages,
    updatePublishedUrl,
    updateFirehoseShared,
    publishedUrl: session.publishedUrl,
  });

  return (
    <div className="flex h-full w-full items-center px-2 py-4">
      <div className="flex w-1/4 items-center justify-start">
        <BackButton
          onBackClick={() => {
            // Tell parent component user explicitly clicked back
            if (isStreaming && setUserClickedBack) {
              setUserClickedBack(true);
            }
            // Force showing the chat panel immediately
            setMobilePreviewShown(false);
          }}
        />

        {showViewControls ? null : <div className="h-10" />}
      </div>

      {/* Center - View controls */}
      <div className="flex w-1/2 items-center justify-center">
        {showViewControls && (
          <ViewControls
            viewControls={viewControls}
            currentView={displayView} // Use displayView for the currently active button highlight
            onClick={navigateToView}
          />
        )}
      </div>
      {/* Right side - Save and Publish buttons */}
      <div className="flex w-1/4 items-center justify-end">
        <div className="flex items-center gap-2">
          {/* Save button - show when in code view and has changes */}
          {displayView === "code" && hasCodeChanges && onCodeSave && (
            <SaveButton
              onClick={onCodeSave}
              hasChanges={hasCodeChanges}
              syntaxErrorCount={syntaxErrorCount}
              testId="header-save-button"
            />
          )}

          {showViewControls &&
            (previewReady ||
              (displayView === "settings" && publishedAppUrl)) && (
              <ShareButton
                ref={publishButtonRef}
                onClick={toggleShareModal}
                isPublishing={isPublishing}
                urlCopied={urlCopied}
                hasPublishedUrl={!!publishedAppUrl}
              />
            )}
        </div>
      </div>
      {/* Share Modal */}
      {isShareModalOpen && (
        <ShareModal
          isOpen={isShareModalOpen}
          onClose={() => setIsShareModalOpen(false)}
          buttonRef={publishButtonRef}
          publishedAppUrl={publishedAppUrl}
          onPublish={handlePublish}
          isPublishing={isPublishing}
          isFirehoseShared={session.firehoseShared}
        />
      )}
    </div>
  );
};

export default ResultPreviewHeaderContent;
