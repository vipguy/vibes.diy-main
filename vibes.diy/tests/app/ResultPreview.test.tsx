import React from "react";
import { vi, describe, it, expect, beforeEach, afterAll } from "vitest";
import { render, screen, act, waitFor } from "@testing-library/react";
import ResultPreview from "~/vibes.diy/app/components/ResultPreview/ResultPreview.js";
import { mockResultPreviewProps } from "./mockData.js";
import { MockThemeProvider } from "./utils/MockThemeProvider.js";

// Mock clipboard API
Object.defineProperty(navigator, "clipboard", {
  value: {
    writeText: vi.fn().mockImplementation(() => Promise.resolve()),
  },
  writable: true,
});

// Mock URL methods that aren't available in test environment
const mockObjectUrl = "mock-blob-url";
URL.createObjectURL = vi.fn().mockReturnValue(mockObjectUrl);
URL.revokeObjectURL = vi.fn();

// Mock useSession hook to prevent Fireproof initialization during tests
vi.mock("~/vibes.diy/app/hooks/useSession", () => ({
  useSession: vi.fn().mockReturnValue({
    updateTitle: vi.fn().mockResolvedValue(undefined),
    session: { title: "Test Session" },
    updatePublishedUrl: vi.fn(),
    updateFirehoseShared: vi.fn(),
    addScreenshot: vi.fn(),
  }),
}));

// Mock SandpackProvider and related components
vi.mock("@codesandbox/sandpack-react", () => ({
  SandpackProvider: ({ children }: { children: React.ReactNode }) => (
    <div data-testid="sandpack-provider">{children}</div>
  ),
  SandpackLayout: ({ children }: { children: React.ReactNode }) => (
    <div>{children}</div>
  ),
  SandpackCodeEditor: () => <div data-testid="sandpack-editor">Editor</div>,
  SandpackPreview: () => <div data-testid="sandpack-preview">Preview</div>,
  useSandpack: () => ({
    sandpack: { activeFile: "/App.jsx" },
    listen: vi.fn().mockReturnValue(() => {
      /* no-op */
    }),
  }),
}));

// Mock WelcomeScreen
vi.mock("~/vibes.diy/app/components/ResultPreview/WelcomeScreen", () => ({
  default: () => <div data-testid="welcome-screen">Welcome Screen Content</div>,
}));

// Mock the Sandpack scroll controller
vi.mock(
  "~/vibes.diy/app/components/ResultPreview/SandpackScrollController",
  () => ({
    default: () => null,
  }),
);

// Mock iframe behavior

// Mock the IframeContent component to avoid iframe issues in tests
vi.mock("~/vibes.diy/app/components/ResultPreview/IframeContent", () => ({
  default: ({ activeView }: { activeView: string }) => (
    <div data-testid="sandpack-provider" className="h-full">
      <div
        style={{
          visibility: activeView === "preview" ? "visible" : "hidden",
          position: activeView === "preview" ? "static" : "absolute",
        }}
      >
        <iframe data-testid="preview-iframe" title="Preview" />
      </div>
      <div
        data-testid="sandpack-editor"
        style={{
          visibility: activeView === "code" ? "visible" : "hidden",
          position: activeView === "code" ? "static" : "absolute",
        }}
      >
        Code Editor Content
      </div>
    </div>
  ),
}));

// Mock ResizeObserver
vi.stubGlobal(
  "ResizeObserver",
  vi.fn().mockImplementation(() => ({
    observe: vi.fn(),
    unobserve: vi.fn(),
    disconnect: vi.fn(),
  })),
);

// Mock window.postMessage for preview communication
const originalPostMessage = window.postMessage;
window.postMessage = vi.fn();

// Original localStorage methods
const originalGetItem = Storage.prototype.getItem;

// Reset mocks between tests
beforeEach(() => {
  globalThis.document.body.innerHTML = "";
  vi.clearAllMocks();
  window.postMessage = vi.fn();

  // Mock localStorage to return a valid API key
  Storage.prototype.getItem = function (key) {
    if (key === "vibes-openrouter-key") {
      return JSON.stringify({ key: "test-api-key", hash: "test-hash" });
    }
    return originalGetItem.call(this, key);
  };
});

// Restore original methods after tests
afterAll(() => {
  window.postMessage = originalPostMessage;
  Storage.prototype.getItem = originalGetItem;
});

describe("ResultPreview", () => {
  it("renders without crashing", () => {
    // Use non-empty code to ensure the editor is shown
    render(
      <MockThemeProvider>
        <ResultPreview
          code="console.log('test');"
          {...mockResultPreviewProps}
        />
      </MockThemeProvider>,
    );

    // Now the sandpack editor should be visible
    expect(screen.getByTestId("sandpack-editor")).toBeInTheDocument();
    // Don't check for preview since it might not be available in the test environment
    // expect(screen.getByTestId('sandpack-preview')).toBeDefined();
  });

  it("displays welcome screen when code is empty", () => {
    const { container } = render(
      <MockThemeProvider>
        <ResultPreview code={""} {...mockResultPreviewProps} />
      </MockThemeProvider>,
    );

    // Instead of finding by role, check that the container has the expected structure
    expect(
      container.querySelector("div.h-full > div.h-full"),
    ).toBeInTheDocument();
  });

  it("handles streaming state correctly", () => {
    const code = 'const test = "Streaming";';

    render(
      <MockThemeProvider>
        <ResultPreview
          code={code}
          isStreaming={true}
          {...mockResultPreviewProps}
        />
      </MockThemeProvider>,
    );

    // Just verify it renders without errors
    expect(screen.getAllByTestId("sandpack-provider")[0]).toBeDefined();
  });

  it("passes dependencies to SandpackProvider", () => {
    const code = 'console.log("test");';
    const dependencies = {
      react: "^18.0.0",
      "react-dom": "^18.0.0",
    };

    render(
      <MockThemeProvider>
        <ResultPreview
          code={code}
          dependencies={dependencies}
          {...mockResultPreviewProps}
        />
      </MockThemeProvider>,
    );

    // Just verify it renders without errors
    expect(screen.getAllByTestId("sandpack-provider")[0]).toBeDefined();
  });

  it("calls onShare when share button is clicked", () => {
    // Skipping test since toolbar with share button has been removed
    // In the future, this would be added to a different component or the header
    expect(true).toBe(true);
  });

  it("shows welcome screen with empty code", () => {
    const { container } = render(
      <MockThemeProvider>
        <ResultPreview code="" {...mockResultPreviewProps} />
      </MockThemeProvider>,
    );

    // Instead of finding by role, check that the container has the expected structure
    expect(
      container.querySelector("div.h-full > div.h-full"),
    ).toBeInTheDocument();
  });

  it("shows a share button when onShare is provided and code is not empty", () => {
    // Skipping test since toolbar with share button has been removed
    expect(true).toBe(true);
  });

  it("updates display when code changes", () => {
    const { rerender } = render(
      <MockThemeProvider>
        <ResultPreview code="" {...mockResultPreviewProps} />
      </MockThemeProvider>,
    );
    rerender(
      <MockThemeProvider>
        <ResultPreview
          code="console.log('test');"
          {...mockResultPreviewProps}
        />
      </MockThemeProvider>,
    );

    // Just verify it renders without errors
    expect(screen.getAllByTestId("sandpack-provider")[0]).toBeDefined();
  });

  it("renders with code content", () => {
    const code = 'const test = "Hello World";';

    render(
      <MockThemeProvider>
        <ResultPreview code={code} {...mockResultPreviewProps} />
      </MockThemeProvider>,
    );

    // Skip button checks since toolbar has been removed
    expect(screen.getByTestId("sandpack-editor")).toBeInTheDocument();
  });

  it("handles copy to clipboard", async () => {
    // Skipping test since toolbar with copy button has been removed
    expect(true).toBe(true);
  });

  it("renders with custom dependencies", async () => {
    const code = 'import React from "react";';
    const dependencies = {
      react: "^18.0.0",
      "react-dom": "^18.0.0",
    };

    render(
      <MockThemeProvider>
        <ResultPreview
          code={code}
          dependencies={dependencies}
          {...mockResultPreviewProps}
        />
      </MockThemeProvider>,
    );

    // Use getAllByTestId to handle multiple elements
    expect(screen.getAllByTestId("sandpack-provider")[0]).toBeInTheDocument();

    // Skip button check since toolbar has been removed
    expect(screen.getByTestId("sandpack-editor")).toBeInTheDocument();
  });

  it("handles share functionality", () => {
    // Skipping test since share button has been removed
    expect(true).toBe(true);
  });

  it("receives preview-ready message from iframe when content loads", async () => {
    // Setup message event listener before rendering
    const previewReadyHandler = vi.fn();
    window.addEventListener("message", (event) => {
      if (event.data?.type === "preview-ready") {
        previewReadyHandler(event.data);
      }
    });

    // Sample code that would be rendered in the iframe
    const code = `
      function App() {
        return <div>Test App Content</div>;
      }
    `;

    const mockSetPreviewLoaded = vi.fn();

    // Create props with our mock onPreviewLoaded, ensuring it overrides the one in mockResultPreviewProps
    const testProps = {
      ...mockResultPreviewProps,
      code,
      isStreaming: false,
      codeReady: true,
      onPreviewLoaded: mockSetPreviewLoaded,
    };

    render(
      <MockThemeProvider>
        <ResultPreview {...testProps} />
      </MockThemeProvider>,
    );

    // Manually trigger the message that would come from the iframe
    const previewReadyEvent = new MessageEvent("message", {
      data: { type: "preview-ready" },
    });

    // Wrap in act() to handle React state updates properly
    act(() => {
      window.dispatchEvent(previewReadyEvent);
    });

    // Wait for the event to be processed
    await waitFor(() => {
      expect(previewReadyHandler).toHaveBeenCalledWith({
        type: "preview-ready",
      });
    });

    // The onPreviewLoaded callback should have been called
    expect(mockSetPreviewLoaded).toHaveBeenCalled();
  });

  it("handles edge case with empty code", () => {
    const { container } = render(
      <MockThemeProvider>
        <ResultPreview code="" {...mockResultPreviewProps} />
      </MockThemeProvider>,
    );

    // Instead of finding by role, check that the container has the expected structure
    expect(
      container.querySelector("div.h-full > div.h-full"),
    ).toBeInTheDocument();
  });

  it("renders empty state correctly", () => {
    const { container } = render(
      <MockThemeProvider>
        <ResultPreview code="" {...mockResultPreviewProps} />
      </MockThemeProvider>,
    );
    // Update snapshot to match new structure
    expect(container).toMatchSnapshot();
  });

  it("handles dependencies correctly", () => {
    const code = `function App() { return <div>Hello World</div>; }`;
    const dependencies = {
      react: "17.0.2",
      "react-dom": "17.0.2",
    };
    render(
      <MockThemeProvider>
        <ResultPreview
          code={code}
          dependencies={dependencies}
          {...mockResultPreviewProps}
        />
      </MockThemeProvider>,
    );

    // Dependencies should be passed to the Sandpack component
    expect(
      screen.queryByText(/Welcome to the preview/i),
    ).not.toBeInTheDocument();
  });

  it("displays code correctly", () => {
    const code = `function App() { return <div>Hello World</div>; }`;
    render(
      <MockThemeProvider>
        <ResultPreview code={code} {...mockResultPreviewProps} />
      </MockThemeProvider>,
    );

    // Code should be processed and displayed in the editor
    expect(
      screen.queryByText(/Welcome to the preview/i),
    ).not.toBeInTheDocument();
  });

  it("shows welcome screen for empty code", () => {
    const { container } = render(
      <MockThemeProvider>
        <ResultPreview code="" {...mockResultPreviewProps} />
      </MockThemeProvider>,
    );

    // Instead of finding by role, check that the container has the expected structure
    expect(
      container.querySelector("div.h-full > div.h-full"),
    ).toBeInTheDocument();
  });

  it("renders code properly", () => {
    render(
      <MockThemeProvider>
        <ResultPreview
          code="console.log('test');"
          {...mockResultPreviewProps}
        />
      </MockThemeProvider>,
    );
    expect(
      screen.queryByText(/Welcome to the preview/i),
    ).not.toBeInTheDocument();
  });

  it("handles code updates correctly", () => {
    const { rerender } = render(
      <MockThemeProvider>
        <ResultPreview code="" {...mockResultPreviewProps} />
      </MockThemeProvider>,
    );
    rerender(
      <MockThemeProvider>
        <ResultPreview
          code="console.log('test');"
          {...mockResultPreviewProps}
        />
      </MockThemeProvider>,
    );

    // Should change from welcome screen to code display
    expect(
      screen.queryByText(/Welcome to the preview/i),
    ).not.toBeInTheDocument();
  });

  it("handles screenshot capture requests", () => {
    const onScreenshotCaptured = vi.fn();
    const code = `function App() { return <div>Hello World</div>; }`;
    render(
      <MockThemeProvider>
        <ResultPreview
          code={code}
          onScreenshotCaptured={onScreenshotCaptured}
          {...mockResultPreviewProps}
        />
      </MockThemeProvider>,
    );

    // Simulate screenshot message
    act(() => {
      window.dispatchEvent(
        new MessageEvent("message", {
          data: { type: "screenshot", data: "base64-data" },
        }),
      );
    });

    expect(onScreenshotCaptured).toHaveBeenCalledWith("base64-data");
  });

  it("handles preview loaded event", async () => {
    const onPreviewLoaded = vi.fn();
    const code = `function App() { return <div>Hello World</div>; }`;
    render(
      <MockThemeProvider>
        <ResultPreview
          code={code}
          {...mockResultPreviewProps}
          onPreviewLoaded={onPreviewLoaded}
        />
      </MockThemeProvider>,
    );

    // Simulate preview loaded message
    act(() => {
      window.dispatchEvent(
        new MessageEvent("message", {
          data: { type: "preview-loaded" },
        }),
      );
    });

    await waitFor(() => {
      expect(onPreviewLoaded).toHaveBeenCalled();
    });
  });

  it("passes dependencies to Sandpack", () => {
    const code = `function App() { return <div>Hello World</div>; }`;
    const dependencies = {
      react: "17.0.2",
      "react-dom": "17.0.2",
    };
    render(
      <MockThemeProvider>
        <ResultPreview
          code={code}
          dependencies={dependencies}
          {...mockResultPreviewProps}
        />
      </MockThemeProvider>,
    );

    // Dependencies should be configured in Sandpack
    expect(
      screen.queryByText(/Welcome to the preview/i),
    ).not.toBeInTheDocument();
  });

  // Test removed: API key functionality has been removed as proxy handles authentication

  it("displays the code editor initially", () => {
    const code = `function App() { return <div>Hello World</div>; }`;
    render(
      <MockThemeProvider>
        <ResultPreview code={code} {...mockResultPreviewProps} />
      </MockThemeProvider>,
    );

    // Should default to code view
    expect(
      screen.queryByText(/Welcome to the preview/i),
    ).not.toBeInTheDocument();
  });

  it("shows welcome screen when no code is provided", () => {
    const { container } = render(
      <MockThemeProvider>
        <ResultPreview code="" {...mockResultPreviewProps} />
      </MockThemeProvider>,
    );

    // Instead of finding by role, check that the container has the expected structure
    expect(
      container.querySelector("div.h-full > div.h-full"),
    ).toBeInTheDocument();
  });

  it("renders with a simple code snippet", () => {
    const code = 'const test = "Hello";';
    // const setActiveView = vi.fn(); // Removed as it's no longer used

    // Render the component with a simple code snippet
    render(
      <MockThemeProvider>
        <ResultPreview
          code={code}
          dependencies={{}}
          isStreaming={false}
          codeReady={true}
          displayView="code" // Changed from activeView
          // setActiveView={setActiveView} // Removed
          onPreviewLoaded={() => {
            /* no-op */
          }}
          setMobilePreviewShown={() => {
            /* no-op */
          }}
          updateTitle={() => Promise.resolve()}
          sessionId="test-session-id"
        />
      </MockThemeProvider>,
    );

    // Now the sandpack editor should be visible
    expect(screen.getByTestId("sandpack-editor")).toBeDefined();
  });
});
