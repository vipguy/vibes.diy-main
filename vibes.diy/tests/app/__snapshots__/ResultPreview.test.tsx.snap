// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ResultPreview > renders empty state correctly 1`] = `
<div>
  <div
    class="h-full"
    style="overflow: hidden; position: relative;"
  >
    <style>
      
  @keyframes spin-slow {
    from {
      transform: rotate(0deg);
    }
    to {
      transform: rotate(360deg);
    }
  }
  .animate-spin-slow {
    animation: spin-slow 1s linear infinite;
  }

    </style>
    <div
      class="h-full"
    />
  </div>
</div>
`;
